src/util/taproot.rs

```diff
@@ -82,12 +93,776 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+        let first_byte: u8 =
+            (if self.output_key_parity { 1 } else { 0 }) | self.leaf_version.as_u8();
+        let mut bytes_written = 0;
+        bytes_written += writer.write(&[first_byte])?;
+        bytes_written += writer.write(&self.internal_key.serialize())?;
+        bytes_written += self.merkle_branch.serialize_to_writer(&mut writer)?;
+        Ok(bytes_written)
+    }
+
+    /// Serialize the control block. This would be required when
+    /// using ControlBlock as a witness element while spending an output via
+    /// script path. This serialization does not include the VarInt prefix that would be
+    /// applied when encoding this element as a witness.
+    pub fn serialize(&self) -> Vec<u8> {
+        let mut buf = Vec::with_capacity(self.size());
+        self.serialize_to_writer(&mut buf)
+            .expect("writers don't error");
+        buf
+    }
+
+    /// Verify that a control block is correct proof for a given output key and script
+    /// This only checks that script is contained inside the taptree described by
+    /// output key, full verification must also execute the script with witness data
+    pub fn verify_taproot_commitment<C: secp256k1::Verification>(
+        &self,
+        secp: &Secp256k1<C>,
+        output_key: &schnorr::PublicKey,
+        script: &Script,
+    ) -> bool {
+        // compute the script hash
+        // Initially the curr_hash is the leaf hash
+        let mut curr_hash = compute_leaf_hash(&script, self.leaf_version);
+        // Verify the proof
+        for elem in self.merkle_branch.as_inner() {
+            let mut eng = TapBranchHash::engine();
+            if &curr_hash < elem {
+                eng.input(&curr_hash);
+                eng.input(elem);
+            } else {
+                eng.input(elem);
+                eng.input(&curr_hash);
+            }
+            // Recalculate the curr hash as parent hash
+            curr_hash = sha256::Hash::from_engine(eng);
+        }
+        // compute the taptweak
+        let tweak = compute_tap_tweak(Some(curr_hash), self.internal_key);
+        self.internal_key.tweak_add_check(
+            secp,
+            output_key,
+            self.output_key_parity,
+            tweak.into_inner(),
+        )
+    }
+}
+
+/// The leaf version for tapleafs
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct LeafVersion {
+    ver: u8,
+}
+
+impl Default for LeafVersion {
+    fn default() -> Self {
+        Self {
+            ver: TAPROOT_LEAF_TAPSCRIPT,
+        }
+    }
+}
+
+impl LeafVersion {
+    /// Obtain LeafVersion from u8, will error when the last bit of ver is not 0
+    pub fn from_u8(ver: u8) -> Result<Self, TaprootError> {
+        if ver & TAPROOT_LEAF_MASK != ver {
+            Err(TaprootError::InvalidTaprootLeafVersion(ver))
+        } else {
+            Ok(LeafVersion { ver: ver })
```

Comment:
This allows construction of an incorrect leaf versions (like 0x50 etc) which will conflict with annex etc. I think we need enumeration of all allowed leaf versions and check against it.

Reply:
here the list of the values from the BIP-341:

> The values that comply to this rule are the 32 even values between 0xc0 and 0xfe and also 0x66, 0x7e, 0x80, 0x84, 0x96, 0x98, 0xba, 0xbc, 0xbe

– <https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#cite_note-7>

Reply:
Good catch about the values. I knew about this rule, but somehow forgot to implement it :(

Reply:
Actually, we only need to test for the annex tag. Other values are only recommendations and not consensus rules.